src/hotspot/os/aix/os_perf_aix.cpp

```diff
@@ -573,267 +436,86 @@ int CPUPerformanceInterface::context_switch_rate(double* rate) const {
     return OS_ERR;
   }
 
-  do {
-    _entry = os::readdir(_dir);
-    if (_entry == NULL) {
-      // Error or reached end.  Could use errno to distinguish those cases.
-      _valid = false;
-      return OS_ERR;
-    }
-  } while(!is_valid_entry(_entry));
-
-  _valid = true;
-  return OS_OK;
-}
-
-SystemProcessInterface::SystemProcesses::ProcessIterator::ProcessIterator() {
-  _dir = NULL;
-  _entry = NULL;
-  _valid = false;
-}
+  records_allocated = *nprocs;
+  proc_stats = NEW_RESOURCE_ARRAY(perfstat_process_t, records_allocated);
 
-bool SystemProcessInterface::SystemProcesses::ProcessIterator::initialize() {
-  // Not yet implemented.
-  return false;
-}
+  // populate stats && set the actual number of procs that have been populated
+  // should never be higher than requested, but may be lower due to process death
+  *nprocs = libperfstat::perfstat_process(&name_holder, proc_stats, sizeof(perfstat_process_t), records_allocated);
 
-SystemProcessInterface::SystemProcesses::ProcessIterator::~ProcessIterator() {
-  if (_dir != NULL) {
-    os::closedir(_dir);
-  }
-}
+  for (int n = 0; n < *nprocs; n++) {
+    psinfo_t psinfo;
+    char* name     = NEW_C_HEAP_ARRAY(char, IDENTIFIER_LENGTH, mtInternal);
+    char* exe_name = NEW_C_HEAP_ARRAY(char, PRFNSZ, mtInternal);
+    char* cmd_line = NEW_C_HEAP_ARRAY(char, PRARGSZ, mtInternal);
 
-SystemProcessInterface::SystemProcesses::SystemProcesses() {
-  _iterator = NULL;
-}
+    strncpy(name, proc_stats[n].proc_name, IDENTIFIER_LENGTH);
 
-bool SystemProcessInterface::SystemProcesses::initialize() {
-  _iterator = new SystemProcessInterface::SystemProcesses::ProcessIterator();
-  return _iterator->initialize();
-}
+    if (read_psinfo(proc_stats[n].pid, psinfo)) {
+      strncpy(exe_name, psinfo.pr_fname, PRFNSZ);
+      strncpy(cmd_line, psinfo.pr_psargs, PRARGSZ);
+    }
 
-SystemProcessInterface::SystemProcesses::~SystemProcesses() {
-  if (_iterator != NULL) {
-    delete _iterator;
+    // create a new SystemProcess with next pointing to current head.
+    SystemProcess* sp = new SystemProcess(proc_stats[n].pid,
+                                          name,
+                                          exe_name,
+                                          cmd_line,
+                                          head);
+    // update head.
+    head = sp;
   }
-}
-
-int SystemProcessInterface::SystemProcesses::system_processes(SystemProcess** system_processes, int* no_of_sys_processes) const {
-  assert(system_processes != NULL, "system_processes pointer is NULL!");
-  assert(no_of_sys_processes != NULL, "system_processes counter pointers is NULL!");
-  assert(_iterator != NULL, "iterator is NULL!");
-
-  // initialize pointers
-  *no_of_sys_processes = 0;
-  *system_processes = NULL;
 
-  while (_iterator->is_valid()) {
-    SystemProcess* tmp = new SystemProcess();
-    _iterator->current(tmp);
+  FREE_RESOURCE_ARRAY(perfstat_process_t, proc_stats, records_allocated);
```

Comment:
no need for release. Use ResourceMark at the top of the function unless RA memory escapes. Or, C heap.